### ISSUE TITLE
FSDP2 tutorial outline

### ISSUE THREAD

**Issue by weifengpy** [pytorch/pytorch#151505]
### 📚 The doc issue

draft the FSDP2 turorial similar to FSDP1's https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html, with code example in [pytorch/examples](https://github.com/pytorch/examples)

basics using [Transformer model](https://github.com/pytorch/pytorch/blob/f5851efed99db3f3509982dda8680c1b60882c6e/torch/testing/_internal/distributed/_tensor/common_dtensor.py#L197)
* model init: nested wrapping, dim-0 sharding, AC
* load state dict: Dtensor version, DCP version
* forward/backward: implicit prefetch and explicit prefetch, reshard_after_forward=False/Int, mixed precision, cpu offloading
* gradient clipping, gradient scaler, optimizer with DTensor
* save state dict: DTensor version, DCP version

advanced topics
* torchrec DMP + DDP + ignored parameters for recommendation models
* HSDP
* tensor subclass extenstion point: float8 example
* dim-i sharding
* composability with TP
* gradient accumulation and composability with PP
* AG/RS buffer in memory pool and sysmetric memory
* inter-stream fragmentation: AG/RS in default pool vs separate pool
* AMD and new accelerator support
* register forward hooks and backward hooks 

FSDP1-to-FSDP2 migration guide
* moving things from https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

**Comment by weifengpy** [comment]
cc @awgu for suggestions on outline

cc @fegin for opinions on load/save state dict with DTensor version, in addition to DCP version

cc @wconstab @wwwjn @tianyu-l @mori360 on outlines

**Comment by awgu** [comment]
I am not sure if it should go in the PyTorch tutorial, but I think a common question for users is: how do I load and save to HuggingFace format?

I think clarity on the loading and saving of state dicts with the DTensor vs. DCP versions is great. When should we _not_ use the DCP version? What minimum version of torch do we need for that? (since I recall there were some bug fixes)

**Comment by joecummings** [comment]
> I am not sure if it should go in the PyTorch tutorial, but I think a common question for users is: how do I load and save to HuggingFace format?
> 
> I think clarity on the loading and saving of state dicts with the DTensor vs. DCP versions is great. When should we _not_ use the DCP version? What minimum version of torch do we need for that? (since I recall there were some bug fixes)

But HuggingFace format do you mean a format that is loadable by the ``from_pretrained`` API in the transformers library or do you generally just mean the safetensors format?

Right now, as I understand it there's no way to do the former without consolidating the state dict and remapping the keys to those expected by the API. This is how we do it in torchtune (for full tensor checkpoints) and it can take a longggg time depending on the size of the model. 

Regarding saving to safetensors, I met with @saumishr yesterday and there is interest in DCP being able to write in a distributed manner to the format.

**Comment by weifengpy** [comment]
> I am not sure if it should go in the PyTorch tutorial, but I think a common question for users is: how do I load and save to HuggingFace format?

good suggestion! torchtune load/save HF format already. I can try minimize it into a clean example. it can be tricky to import HF in pytorch example, but I will find a place to host the code - maybe torchtitan or simply gitst

**Comment by weifengpy** [comment]
for loading HF checkpoint, we can reuse code from https://github.com/pytorch/pytorch/issues/151936 and https://github.com/pytorch/torchtitan/issues/1133 to facilitate
